File: 2025-1/sesion-de-clase-18/SoftProgNet/SoftProgPersistencia/DAOImpl/RRHH/EmpleadoDAOImpl.cs
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PUCP.Edu.Pe.SoftProg.Modelo.RRHH;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAO.RRHH;
using PUCP.Edu.Pe.SoftProg.Config;

namespace PUCP.Edu.Pe.SoftProg.Persistencia.DAOImpl.RRHH {
    public class EmpleadoDAOImpl : BaseDAOImpl<Empleado>, IEmpleadoDAO {
        protected override DbCommand CommandoInsertar(DbConnection conn, Empleado empleado) {

            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarEmpleado";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idArea", DbType.Int32, empleado.Area.Id);

            if (empleado.CuentaUsuario != null) {
                this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, empleado.CuentaUsuario.Id);
            }
            else {
                this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, DBNull.Value);
            }

            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, empleado.Dni);

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, empleado.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, empleado.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.String, empleado.Genero);

            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, empleado.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_cargo", DbType.String, empleado.Cargo);
            this.AgregarParametroEntrada(cmd, "@p_sueldo", DbType.Double, empleado.Sueldo);

            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, empleado.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand CommandoInsertar(DbConnection conn, DbTransaction transaccion, Empleado empleado) {
            DbCommand cmd = this.CommandoInsertar(conn, empleado);
            cmd.Transaction = transaccion;
            return cmd;
        }

        protected override DbCommand CommandoModificar(DbConnection conn, Empleado empleado) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarEmpleado";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idArea", DbType.Int32, empleado.Area.Id);
            this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, empleado.Dni);

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, empleado.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, empleado.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.String, empleado.Genero);

            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, empleado.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_cargo", DbType.String, empleado.Cargo);
            this.AgregarParametroEntrada(cmd, "@p_sueldo", DbType.Double, empleado.Sueldo);

            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, empleado.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, empleado.Id);

            return cmd;
        }

        protected override DbCommand CommandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarEmpleado";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoBuscar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarEmpleadoPorId";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoListar(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarEmpleados";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            return cmd;
        }

        protected override Empleado mapearModelo(DbDataReader reader) {
            return new Empleado {
                Id = Convert.ToInt32(reader["id"]),
                Dni = Convert.ToString(reader["dni"]), 
                Nombre = Convert.ToString(reader["nombre"]),
                ApellidoPaterno = Convert.ToString(reader["apellidoPaterno"]), 
                Genero = Convert.ToChar(reader["genero"]),
                FechaNacimiento = Convert.ToDateTime(reader["fechaNacimiento"]),
                Cargo = Convert.ToString(reader["cargo"]), 
                Sueldo = Convert.ToDouble(reader["sueldo"]), 
                IsActive = (bool)reader["activo"],
                Area = new AreaDAOImpl().Buscar(Convert.ToInt32(reader["idArea"]))
            };
        }

        public List<Empleado> BuscarPorDni(string dni) {
            try {
                using (DbConnection conn = DBManager.GetInstance().GetConnection())
                using (DbCommand cmd = this.ComandoBuscarPorDni(conn, dni))
                using (DbDataReader reader = cmd.ExecuteReader()) {
                    List<Empleado> modelos = new List<Empleado>();

                    while (reader.Read()) {
                        modelos.Add(this.mapearModelo(reader));
                    }

                    return modelos;
                }
            }
            catch (Exception e) {
                Console.Error.WriteLine("Error SQL Durante la busueda: " + e.Message);
                throw new InvalidOperationException("No se pudo buscar el registro.", e);
            }
        }

        private DbCommand ComandoBuscarPorDni(DbConnection conn, string dni) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarEmpleadoPorDni";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.Int32, dni);

            return cmd;
        }
    }
}

File: 2025-1/sesion-de-clase-20/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/rrhh/EmpleadoDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.rrhh;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.Date;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.config.DBManager;
import pe.edu.pucp.inf30.softprog.dao.rrhh.IEmpleadoDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAOImpl;
import pe.edu.pucp.inf30.softprog.model.rrhh.Empleado;

public class EmpleadoDAOImpl extends BaseDAOImpl<Empleado> implements IEmpleadoDAO {    
    @Override
    protected CallableStatement comandoInsertar(Connection conn, Empleado empleado) throws SQLException {
        String sql = "{CALL insertarEmpleado(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);   
        cmd.setInt("p_idArea", empleado.getArea().getId());
        
        if (empleado.getCuentaUsuario() != null) {
            cmd.setInt("p_idCuentaUsuario", empleado.getCuentaUsuario().getId());
        }
        else {
            cmd.setNull("p_idCuentaUsuario", Types.INTEGER);
        }
        
        cmd.setString("p_dni", empleado.getDni());
        cmd.setString("p_nombre", empleado.getNombre());
        cmd.setString("p_apellidoPaterno", empleado.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(empleado.getGenero()));
        cmd.setDate("p_fechaNacimiento", new Date(empleado.getFechaNacimiento().getTime()));
        cmd.setString("p_cargo", empleado.getCargo());
        cmd.setDouble("p_sueldo", empleado.getSueldo());
        cmd.setBoolean("p_activo", empleado.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        return cmd;
    }
    
    @Override
    protected CallableStatement comandoModificar(Connection conn, Empleado empleado) throws SQLException {
        String sql = "{CALL modificarEmpleado(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idArea", empleado.getArea().getId());
        
        if (empleado.getCuentaUsuario() != null) {
            cmd.setInt("p_idCuentaUsuario", empleado.getCuentaUsuario().getId());
        }
        else {
            cmd.setNull("p_idCuentaUsuario", Types.INTEGER);
        }
        
        cmd.setString("p_dni", empleado.getDni());
        cmd.setString("p_nombre", empleado.getNombre());
        cmd.setString("p_apellidoPaterno", empleado.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(empleado.getGenero()));
        cmd.setDate("p_fechaNacimiento", new Date(empleado.getFechaNacimiento().getTime()));
        cmd.setString("p_cargo", empleado.getCargo());
        cmd.setDouble("p_sueldo", empleado.getSueldo());
        cmd.setBoolean("p_activo", empleado.isActivo());
        cmd.setInt("p_id", empleado.getId());
        return cmd;
    }

    @Override
    protected CallableStatement comandoEliminar(Connection conn, int id) throws SQLException {
        String sql = "{CALL eliminarEmpleado(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoBuscar(Connection conn, int id) throws SQLException {
        String sql = "{CALL buscarEmpleadoPorId(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoListar(Connection conn) throws SQLException {
        String sql = "{CALL listarEmpleados()}";
        CallableStatement cmd = conn.prepareCall(sql);
        return cmd;
    }

    @Override
    protected Empleado mapearModelo(ResultSet rs) throws SQLException {
        Empleado empleado = new Empleado();
        empleado.setId(rs.getInt("id"));
        empleado.setArea(new AreaDAOImpl().buscar(rs.getInt("idArea")));
        //empleado.setCuentaUsuario(new CuentaUsuarioDAOImpl().buscar(rs.getInt("idCuentaUsuario")));
        empleado.setDni(rs.getString("dni"));
        empleado.setNombre(rs.getString("nombre"));
        empleado.setApellidoPaterno(rs.getString("apellidoPaterno"));
        empleado.setGenero(rs.getString("genero").charAt(0));
        empleado.setFechaNacimiento(rs.getTimestamp("fechaNacimiento"));
        empleado.setCargo(rs.getString("cargo"));
        empleado.setSueldo(rs.getDouble("sueldo"));
        empleado.setActivo(rs.getBoolean("activo"));

        return empleado;
    }

    protected CallableStatement comandoBuscarPorDni(Connection conn, String dni) throws SQLException {
        String sql = "{CALL buscarEmpleadoPorDni(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_dni", dni);
        return cmd;
    }
    
    @Override
    public Empleado buscarPorDni(String dni) {
        try (
            Connection conn = DBManager.getInstance().getConnection();
            CallableStatement cmd = this.comandoBuscarPorDni(conn, dni);
        ) {
            ResultSet rs = cmd.executeQuery();
            
            if (!rs.next()) {
                System.err.println("No se encontro el registro con dni: " + dni);
                return null;
            }
            
            return this.mapearModelo(rs);
        }
        catch (SQLException e) {
            System.err.println("Error SQL durante la insercion: " + e.getMessage());
            throw new RuntimeException("No se pudo insertar el registro.", e);
        }
        catch (Exception e) {
            System.err.println("Error inpesperado: " + e.getMessage());
            throw new RuntimeException("Error inesperado al insertar el registro.", e);
        }
    }
}
